Provide attribute macro for multiple HTTP methods

mattgathu opened the pull request:
## PR Type
Feature

## PR Checklist
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt


## Overview
Provide attribute macro for multiple HTTP methods

Define a new `route` attribute macro that supports defining multiple
HTTP methods to be routed to (handled by) a single handler.

The attribute macro syntax looks like this
```rust
use actix_web::route;

#[route("/multi", method = "GET", method = "POST")]
async fn multi_methods() -> &'static str {
    "Hello world!\r\n"
}
```

This implementation extends the [`GuardType`][1] enum in actix-web-codegen to have a
new `GuardType::Multi` variant that denotes when multiple method guards
are used.

A new `method` attribute in the `route` attribute macro provides an  HTTP method to provide a guard for.
This attribute can be used multiple times for multiple HTTP methods.

The code matches the methods to the respective
`GuardType` and uses the [`AnyGuard`][2] struct to combine them together.

The generated code looks like this:
```rust
pub struct multi_methods;
impl actix_web::dev::HttpServiceFactory for multi_methods {
    fn register(self, __config: &mut actix_web::dev::AppService) {
        async fn multi_methods() -> &'static str {
            "Hello world!\r\n"
        }
        let __resource = actix_web::Resource::new("/multi")
            .name("multi_methods")
            .guard(actix_web::guard::Any(actix_web::guard::Get()).or(actix_web::guard::Post()))
            .to(multi_methods);
        actix_web::dev::HttpServiceFactory::register(__resource, __config)
    }
}
```

<!-- If this PR fixes or closes an issue, reference it here. -->
<!-- Closes #000 -->
This Closes https://github.com/actix/actix-web/issues/1360

[1]: https://github.com/actix/actix-web/blob/master/actix-web-codegen/src/route.rs#L21
[2]: https://github.com/actix/actix-web/blob/master/src/guard.rs#L104s

<!-- Thanks for considering contributing actix! -->
<!-- Please fill out the following to make our reviews easy. -->

mattgathu committed:
Provide attribute macro for multiple HTTP methods

What
--
Define a new `route` attribute macro that supports defining multiple
HTTP methods to routed to (handled by) a single handler.

The attribute macro syntax looks like this
```rust
use actix_web::route;

async fn multi_methods() -> &'static str {
    "Hello world!\r\n"
}
```

How
--
This implementation extends the [`GuardType`][1] enum in actix-web-codegen to have a
new `GuardType::Multi` variant that denotes when multiple method guards
are used.

A new `methods` attribute in the `route` attribute macro provides a
comma-separated list of HTTP methods to provide guard for.

The code parses the methods list, matches them to the respective
`GuardType` and uses the `AnyGuard` struct to combine them together.

A constructor method for [`AnyGuard`][2] is added to support this.

The generated code looks like this:
```rust
pub struct multi_methods;
impl actix_web::dev::HttpServiceFactory for multi_methods {
    fn register(self, __config: &mut actix_web::dev::AppService) {
    ¦   async fn multi_methods() -> &'static str {
    ¦   ¦   "Hello world!\r\n"
    ¦   }
    ¦   let __resource = actix_web::Resource::new("/multi")
    ¦   ¦   .name("multi_methods")
    ¦   ¦   .guard(actix_web::guard::AnyGuard::new(<[_]>::into_vec(box [
    ¦   ¦   ¦   Box::new(actix_web::guard::Get()),
    ¦   ¦   ¦   Box::new(actix_web::guard::Post()),
    ¦   ¦   ])))
    ¦   ¦   .to(multi_methods);
    ¦   actix_web::dev::HttpServiceFactory::register(__resource, __config)
    }
}
```

**NOTE: This is my first attempt that implementing this feature.
Feedback and mentorship is highly welcome to improve it :-)**

Why
--
This fixes https://github.com/actix/actix-web/issues/1360

[1]: https://github.com/actix/actix-web/blob/master/actix-web-codegen/src/route.rs#L21
[2]: https://github.com/actix/actix-web/blob/master/src/guard.rs#L104s

mattgathu committed:
Add route macro tests

Silentdoer commented:
I think this api should be:
```rust
#[route("/multi", method = "GET", method = "POST")]
async fn multi_methods() -> &'static str {
    "Hello world!\r\n"
}
```

otherwise it's not consistent with:
```rust
#[get("/index2", wrap = "middleware::SayHi", wrap = "middleware::SayHi2")]
```

Silentdoer commented:
or may be:
```rust
#[route("/multi", methods = ["GET","POST"])]
async fn multi_methods() -> &'static str {
    "Hello world!\r\n"
}
```

mattgathu commented:
Thanks for the feedback @Silentdoer 

My initial approach was `#[route("/multi", methods = ["GET","POST"])]` but unfortunately only literal arguments are allowed in this position, and that ruled out the `[...]` syntax.

I can reimplement the macro to have the `#[route("/multi", method = "GET", method = "POST")]` syntax. If the core contributor(s) can affirm that this is the preferred syntax. maybe @robjtede or @JohnTitor

mattgathu committed:
Update route attribute macro syntax

mattgathu commented:
I've updated the syntax to be `#[route("/multi", method = "GET", method = "POST")]`

mattgathu committed:
Update Change log

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -96,10 +100,38 @@ impl Args {
                         }
+                    } else if nv.path.is_ident("method") {
+                        if let syn::Lit::Str(ref lit) = nv.lit {
+                            match lit.value().to_uppercase().as_str() {
```
I don't think that `to_uppercase()` is required. HTTP methods are commonly written in upper case and supporting values such as "cOnNeCt" or "Get" may be confusing.

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -57,13 +59,15 @@ struct Args {
         let mut path = None;
         let mut guards = Vec::new();
         let mut wrappers = Vec::new();
+        let mut methods = Vec::new();
```
What will happen if user will provide multiple methods with the same name? E.g. `method = "POST", method = "POST"`.

mattgathu replied:
we would register an extra redundant guard. This can be avoided by making the methods vec a set or returning a compile time error. _not sure which is the better approach_

popzxc replied:
I would prefer using a set AND returning an error if method is met more than once. Luckily, `HashSet::insert` returns a bool indicating whether this method was already present inside.

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -166,6 +199,13 @@ impl Route {
+        if guard == GuardType::Multi && args.methods.is_empty() {
+            return Err(syn::Error::new(
+                Span::call_site(),
+                "The #[route(..)] macro requires at least one `method` attribute!",
```
Let's keep ourselves calm 🙃 

```suggestion
                "The #[route(..)] macro requires at least one `method` attribute",
```

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -201,25 +241,47 @@ impl ToTokens for Route {
-                        .#resource_type(#name);
-
-                    actix_web::dev::HttpServiceFactory::register(__resource, __config)
+        let stream = if guard != &GuardType::Multi {
```
`if` branches differ in one line only. I suggest generating only that line inside `if`, and then use it as a part of the shared `quote`, like:

```rust
let guard = if guard != &GuardType::Multi {
    // ...
} else {
    // ...
};

let stream = quote! { ... };
```

mattgathu committed:
Simplify code generation + remove uppercase conversion

* assume HTTP methods always specified in uppercase e.g `GET`
* simplify code generation by quote!

mattgathu committed:
Handle repeated HTTP Method edge case

Refactor code to detect when the same HTTP method is specified more than
once. eg. `#[route("/multi", method = "GET", method = "GET")]`
and return a compile time error.

popzxc commented on actix-web-codegen/src/lib.rs:
```diff
@@ -141,6 +141,20 @@ pub fn patch(args: TokenStream, input: TokenStream) -> TokenStream {
+/// Syntax: `#[route("path"[, attributes])]`
+///
+/// ## Attributes
+/// - `"path"` - Raw literal string with path for which to register handler. Mandatory.
```
```suggestion
/// ## Attributes
///
/// - `"path"` - Raw literal string with path for which to register handler. Mandatory.
```

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -201,10 +255,23 @@ impl ToTokens for Route {
         let resource_name = name.to_string();
+        let methods = methods.iter();
+
+        let guard_gen = if guard == &GuardType::Multi {
```
For a one-byte types it's better to derive `Copy` and compare them by value, rather than using references.

mattgathu replied:
I guess I can do an deref `let guard_gen = if *guard == GuardType::Multi`?

popzxc replied:
Isn't it possible to destruct a `Copy` field by value from a reference? If not, then yep, deref seems right.

mattgathu replied:
I still need to a deref even when I derive` Copy` and `Clone` for `GuardType`. I'm guessing the compiler doesn't treat enums as primitives types such as i32 or u8 ¯\_(ツ)_/¯

popzxc commented on actix-web-codegen/src/route.rs:
```diff
@@ -96,10 +102,50 @@ impl Args {
                             ));
                         }
+                    } else if nv.path.is_ident("method") {
+                        if let syn::Lit::Str(ref lit) = nv.lit {
```
Would it make sense to provide an implementation of `TryFrom<&syn::Lit> for GuardType`?

With that, the code here will be simplified to just:

```rust
let guard = GuardType::try_from(&nv.lit)?;
if !methods.insert(guard) { ... };
```

mattgathu replied:
I managed this:
```rust
                    } else if nv.path.is_ident("method") {
                        if let syn::Lit::Str(ref lit) = nv.lit {
                            let guard = GuardType::try_from(lit)?;
                            if !methods.insert(guard) {
                                return Err(syn::Error::new_spanned(
                                    &nv.lit,
                                    &format!(
                                        "HTTP Method defined more than once: `{}`",
                                        lit.value()
                                    ),
                                ));
                            }
                        } else {
                            return Err(syn::Error::new_spanned(
                                nv.lit,
                                "Attribute method expects literal string!",
                            ));
                        }
                    } else {
``` 
So doing the conversation from `&syn::LitStr` because using `&syn::Lit` would mean having the `"Attribute method expects literal string!"` error inside the `GuardType` conversion which seems wrong.

popzxc replied:
Makes sense.

mattgathu committed:
Implement  TryFrom<&syn::LitStr> for GuardType

Try converting String literal to GuardType

robjtede commented:
Hey @mattgathu, it's great to see you tackling this!

I'd like to land #1677 first so we can test some failure cases in CI and make sure the compile error outputs make sense.

popzxc reviewed (approved):
Looks really nice. Thanks!

robjtede commented:
@mattgathu #1677 is merged so it'd be great to get some trybuild tests added as part of this. Try to get in as many weird edge case failures as you can think of.

mattgathu committed:
Add trybuild tests for route macro

robjtede commented on src/guard.rs:
```diff
@@ -104,6 +104,10 @@ pub fn Any<F: Guard + 'static>(guard: F) -> AnyGuard {
+    /// Create AnyGuard from a vector of Guards
+    pub fn new(guards: Vec<Box<dyn Guard>>) -> Self {
+        AnyGuard(guards)
+    }
```
Seems to me this change is unnecessary, it should be possible to use `Any(first_guard).or(second_guard)...`.

mattgathu replied:
I think it is necessary since there is no way (_at least for me_) to repeatedly call `Any(first_guard).or(second_guard)` when doing code generation using `quote!` in `actix-web-codegen`

mattgathu commented:
@robjtede I got an issue with tests passing on **rustc 1.42.0** but failing on the **current stable**. 
Here is the diff:
```
EXPECTED:
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
error: The #[route(..)] macro requires at least one `method` attribute
##[error] --> $DIR/route-missing-method-fail.rs:3:1
  |
3 | #[route("/")]
  | ^^^^^^^^^^^^^

error[E0425]: cannot find value `index` in this scope
##[error]  --> $DIR/route-missing-method-fail.rs:10:49
   |
10 |     let srv = test::start(|| App::new().service(index));
   |                                                 ^^^^^ not found in this scope
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
ACTUAL OUTPUT:
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
error: The #[route(..)] macro requires at least one `method` attribute
##[error] --> $DIR/route-missing-method-fail.rs:3:1
  |
3 | #[route("/")]
  | ^^^^^^^^^^^^^
  |
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `index` in this scope
##[error]  --> $DIR/route-missing-method-fail.rs:10:49
   |
10 |     let srv = test::start(|| App::new().service(index));
   |                                                 ^^^^^ not found in this scope
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
note: If the actual output is the correct output you can bless it by rerunning
      your test with the environment variable TRYBUILD=overwrite

```
note the **= note: this error originates in an attribute macro...** line.

Do you have any suggestions on how to handle this?

robjtede commented:
> Do you have any suggestions on how to handle this?

Hmm... I'll have a look.

In the mean time, I feel that the macro should be called "resource" as it is more consistent with the manual routing methods. A resource can define multiple methods on a path where a route is a single method.

robjtede commented on actix-web-codegen/src/lib.rs:
```diff
@@ -141,6 +141,21 @@ pub fn patch(args: TokenStream, input: TokenStream) -> TokenStream {
     route::generate(args, input, route::GuardType::Patch)
 }
 
+/// Creates route handler with Multiple HTTP methods guards.
```
```suggestion
/// Creates resource handler, allowing multiple HTTP method guards.
```

robjtede commented on actix-web-codegen/src/lib.rs:
```diff
@@ -141,6 +141,21 @@ pub fn patch(args: TokenStream, input: TokenStream) -> TokenStream {
 
+/// Creates route handler with Multiple HTTP methods guards.
+///
+/// Syntax: `#[route("path"[, attributes])]`
```
Part of this doc block should clearly show that this macro is designed to be used with multiple methods.

mattgathu replied:
I've added an `Example` line, similar to the `Syntax` line.

mattgathu commented:
> In the mean time, I feel that the macro should be called "resource" as it is more consistent with the manual routing methods. A resource can define multiple methods on a path where a route is a single method.

I would argue that `route` is more fitting as it expresses the intention: to route one or more http methods to the same handler.  It is also, I would argue, more consistent with the HTTP verbs `get`, `put` etc as it is also a *verb*.

`resource` is a noun and doesn't capture/express the user's intent. 

_I'm viewing `route` as a verb and not its noun meaning_

mattgathu committed:
Update actix-web-codegen/src/lib.rs

Co-authored-by: Rob Ede <robjtede@icloud.com>

robjtede commented:
Alright, fair argument. Would like to get more opinons from @actix/contributors then.

When you read it as a verb it makes sense even though you'd have to sort of read it as `#[route_to("/abc")]`. I think there's value in keeping the nomenclature consitent with existing routing methods. But, again, more input here will be good.

mattgathu committed:
Add example in docs to illustrate route macro use

mattgathu committed:
Use rustversion to pin tests to rustc 1.42

mattgathu committed:
Fix rustversion annotation

robjtede commented:
I'm coming around to #[route(...)] for brevity sake.

In other news, I think I've found a solution to the compile error difference while letting us test on stable too. Great find with that rustversion macro; using that you can do this:

```rust
#[test]
fn compile_macros() {
    let t = trybuild::TestCases::new();

    t.pass("tests/trybuild/simple.rs");
    t.compile_fail("tests/trybuild/simple-fail.rs");

    t.pass("tests/trybuild/route-ok.rs");
    t.compile_fail("tests/trybuild/route-duplicate-method-fail.rs");
    t.compile_fail("tests/trybuild/route-unexpected-method-fail.rs");
    missing_method_fail_hack(&t);
}

#[rustversion::stable(1.42)]
fn missing_method_fail_hack(t: &trybuild::TestCases) {
    t.compile_fail("tests/trybuild/route-missing-method-fail-msrv.rs");
}

#[rustversion::not(stable(1.42))]
fn missing_method_fail_hack(t: &trybuild::TestCases) {
    t.compile_fail("tests/trybuild/route-missing-method-fail.rs");
}
```

Then have two versions of the test file...:
- route-missing-method-fail.rs
- route-missing-method-fail-msrv.rs

(Even cleaner, you can symlink the -msrv version to the original.)

...and two versions of the compiler output:
- route-missing-method-fail.stderr
- route-missing-method-fail-msrv.stderr

fakeshadow commented:
I'm not against this poll but would like to share my two cents.

I feel attributes should follow the rust std style. For example: 
#[route(method = "")]
#[route(any(method = "", method = "" ))]
#[route(not(method = ""))]

mattgathu commented:
@robjtede regarding the trybuild tests, the Actix-web's CI matrix has 1.42, current stable and nightly tests with the compilers never guaranteed to produce similar outputs. There will be a probability that each time nightly or the current stable is updated the tests will break. Pinning the tests to 1.42 avoids this likely breakage.

Having multiple tests for different compilers is great but might not be maintainable IMHO.

robjtede commented:
I think the maintenance cost is small enough using this approach, testing for MSRV and latest stable is enough; we should not run any trybuild tests on nightly.

Worst case we will need to update the stderr files every 6 weeks. But I think theres value in knowing when compiler output changes for our macro errors, anyway.

mattgathu commented:
@robjtede I haven't found a way to differentiate between the 1.42 stable and other stable versions using the rustversion macros :-( 

You have `#[rustversion::not(stable(1.42))]` in your example, but it doesn't eliminate nightly runs.

**Edit:** I ended up building a dummy shim and stacking the macros 🤞

mattgathu committed:
Support stable versions in tests

robjtede commented:
Great! I'm happy to go ahead with that test set up.

Check out this commit (02e16c6538ae31351e7618a714d42180046e4e90) for how to do the multi guard without requiring changes to actix-web. The generated code looks like this for the example in that commit:

```rust
let __resource = actix_web::Resource::new("/")
    .name("index")
    .guard(
        actix_web::guard::Any(actix_web::guard::Get())
            .or(actix_web::guard::Trace())
            .or(actix_web::guard::Post())
            .or(actix_web::guard::Head()),
    )
    .to(index);
```

mattgathu committed:
Implement methods guards using the .or syntax

robjtede reviewed (approved):
Thanks so much for working on this 👍 great job!

robjtede commented:
@fakeshadow RE your comment on attribute style: I don't think there's a use case for not(...) here and I think it's safe to assume the syntax is any(...) without having to explicitly write that out. This is also inkeeping with current wrap="..." style.

Speak now if you still feel strongly on it.

fakeshadow commented:
I'm fine with the current attributes. You guys did a good job.

robjtede commented:
[And awaaayyyyy we go...](https://media.giphy.com/media/WOYKaXG2xJsBO/giphy.gif)
